### DISCUSSION TITLE
[FAQ] How do i delete all mails except a single domain?

### DISCUSSION THREAD

**Discussion by Smig0l** [jsuto/piler#47] [Q&A] [answered]
I need to delete all the mail excluding @domain1.com.
In my case i am archiving @domain1.com and @domain2.com mails.
So according to the wiki i need to define a retention policy (configuring FROM/TO @domain2.com) or select and "delete" manually using the auditor.
The issue is that if for example @domain1 have sent emails to @domain2 these will be deleted using the retention policy...

piler version: 1.3.5 build 997

EDIT: so in my case I want to delete all,  excluding mail that have @domain1.com either in the FROM or TO field

**Comment by Smig0l** [comment]
what about deleting the records from the db? is it enough?

**Reply by jsuto** [reply]
Well, no. The database contains only the metadata. The actual email files are still on the disk. Try using the v_messages view that contains both the sender and the recipients. I'll write another comment soon to describe how.

**Reply by jsuto** [reply, marked as answer]
Try the following (be sure to backup the piler database and/or perhaps the whole archive before starting).

Set the retention column for all messages to the past if they are not sent to or received from "domain1", eg. 

```
update metadata set retained=1000000 where id not in (select id from v_messages where fromdomain='domain1' or todomain='domain1');
```

Then run pilerpurge.py in dry mode, eg.

```
/usr/libexec/piler/pilerpurge.py -d
```

It prints a bunch of files and IDs what it would remove from the system. eg.

```
[1, 2, 18, 20, 21, 22, 23, 39, 40, 41, 42, 43, 44, 45, 46, 47, 49, 50, 51, .... 418, 420]
```

You may check these IDs in the v_messages view in the piler database to ensure that they are NOT from/to domain1 that you want to preserve, eg.

```
select id, fromdomain, todomain from v_messages where id in (1, 2, 18, 20, 21, 22, 23, 39, 40, 41, 42, 43, 44, 45, 46, 47, 49, 50, 51, .... 418, 420);
```

Once you are confident, run pilerpurge.py as user piler to get rid of the unwanted emails:

```/usr/libexec/piler/pilerpurge.py```

**Comment by jsuto** [comment]
Add the following to piler.conf:

```
queuedir=/var/piler/store
```

**Comment by Smig0l** [comment]
so i executed the query and ran the `/usr/local/libexec/piler/pilerpurge.py --config /usr/local/etc/piler/piler.conf` using the -d option.
it was running and outputted a bunch of ids... i killed it and then tried without the dry-run but now it just close itself after a bunch of seconds.
i checked under `/var/piler/store` and it seems that some files were created under it when i used the dry-run.
any hints?

**Reply by jsuto** [reply]
pilerpurge doesn't create any files. However, it syslogs what it does, so check the mail log to figure out what happened.

**Reply by Smig0l** [reply]
>so check the mail log to figure out what happened

`localhost pilerpurge.py[24067]: Purged 0 messages, 0 files, 0/0 bytes`

strange because i updated the messages with retained 1000000...and some ids were printed using dry-run

**Reply by jsuto** [reply]
It's odd. Run the following mysql query to check if pilerpurge has removed these messages:

```select id, deleted from metadata where retained=1000000 limit 10;```

**Reply by Smig0l** [reply]
> It's odd. Run the following mysql query to check if pilerpurge has removed these messages:
> 
> `select id, deleted from metadata where retained=1000000 limit 10;`

they are still there...now even using the dry-run does nothing...
they are marked as deleted = 1....

**Reply by jsuto** [reply]
The pilerpurge utility does not remove the database entries, only the files from the /var/piler/store dirs.
Try the following:

```select id, deleted, piler_id from metadata where retained=1000000 limit 10;```

Then look for files in /var/piler/store/ subdirs like <piler_id>.m, eg. 400000000abcdef378373.m
If you can find them then they haven't been removed.

**Reply by Smig0l** [reply]
thanks! they seems to have been deleted correctly.
